### PR TITLE
TransactionView: AddressTableLookupMeta

### PR DESCRIPTION
**Pull request by apfitzge**
#### Problem
- Task in #2255

#### Summary of Changes
- Parse `MessageAddressTableLookup`s

Fixes #

### CODE REVIEW

File: transaction-view/src/address_table_lookup_meta.rs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        bytes::{
+            advance_offset_for_array, advance_offset_for_type, check_remaining,
+            optimized_read_compressed_u16, read_byte,
+        },
+        result::Result,
+    },
+    solana_sdk::{packet::PACKET_DATA_SIZE, pubkey::Pubkey},
+};
+
+/// Contains meta-data about the address table lookups in a transaction packet.
```
Comment (LucasSte) [inline transaction-view/src/address_table_lookup_meta.rs]:
```suggestion
/// Contains metadata about the address table lookups in a transaction packet.
```

File: transaction-view/src/address_table_lookup_meta.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    /// Get the number of address table lookups (ATL) and offset to the first.
+    /// The offset will be updated to point to the first byte after the last
+    /// ATL.
+    /// This function will parse each ATL to ensure the data is well-formed,
+    /// but will not cache data related to these ATLs.
+    pub fn try_new(bytes: &[u8], offset: &mut usize) -> Result<Self> {
+        // Read the number of ATLs at the current offset.
+        // Each ATL needs at least 34 bytes, so do a sanity check here to
+        // ensure we have enough bytes to read the number of ATLs.
+        const MIN_SIZED_ATL: usize = core::mem::size_of::<Pubkey>() + 1 + 1;
```
Comment (LucasSte) [inline transaction-view/src/address_table_lookup_meta.rs]:
```suggestion
        const MIN_SIZED_ATL: usize = core::mem::size_of::<Pubkey>() + 2;
```

File: transaction-view/src/address_table_lookup_meta.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        const MAX_ATLS_PER_PACKET: usize = PACKET_DATA_SIZE / MIN_SIZED_ATL;
+        // Maximum number of ATLs should be represented by a single byte,
+        // thus the MSB should not be set.
+        const _: () = assert!(MAX_ATLS_PER_PACKET & 0b1000_0000 == 0);
```
Comment (LucasSte) [inline transaction-view/src/address_table_lookup_meta.rs:63]: I forgot about this for your other PRs, but Rust has `const_assert_eq` and `const_assert` (already available in our monorepo) that check constraints during compile time. Perhaps you can use them here (and in all the other cases - feel free to do that in another PR if you like my idea).

File: transaction-view/src/address_table_lookup_meta.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        let num_address_table_lookups = u16::from(read_byte(bytes, offset)?);
```
Comment (LucasSte) [inline transaction-view/src/address_table_lookup_meta.rs]: The number is a `u8`. Why did you choose `u16` to save it?

Comment (apfitzge) [reply transaction-view/src/address_table_lookup_meta.rs]: was consistently using u16s for all my sizes; but you're right. I should just use u8s whenever those cannot validly exceed a u8.

File: transaction-view/src/address_table_lookup_meta.rs (hunk 1 above)
Comment (apfitzge) [reply transaction-view/src/address_table_lookup_meta.rs]: cdbd735dbea1978288ab6fbfa1e61c7ee1ba2f8d

File: transaction-view/src/address_table_lookup_meta.rs (hunk 2 above)
Comment (apfitzge) [reply transaction-view/src/address_table_lookup_meta.rs]: 4d5be57b3f4f0cde9cad4b6f6cd67146daa57c47

I added comments to each number to clarify what it is representing.

File: transaction-view/src/address_table_lookup_meta.rs (hunk 4 above)
Comment (apfitzge) [reply transaction-view/src/address_table_lookup_meta.rs]: c0d298dc65c7fb06d5d58caca17fc32467da77dc

File: transaction-view/src/address_table_lookup_meta.rs (hunk 3 above)
Comment (apfitzge) [reply transaction-view/src/address_table_lookup_meta.rs:63]: Haven't looked at the crate that provides those, but have been trying to keep dependencies to minimum.

Doesn't seem necessary except for slightly nicer syntax, but I don't feel the syntax here is that bad honestly. Maybe you/others feel differently?

Comment (LucasSte) [reply transaction-view/src/address_table_lookup_meta.rs:63]: The nicer syntax is my preference, but this is also a nitpick. I'll leave it up to you. Perhaps, just ask a second opinion from the other reviewers.

File: transaction-view/src/address_table_lookup_meta.rs (hunk 4 above)
Comment (apfitzge) [reply transaction-view/src/address_table_lookup_meta.rs]: After my 2 open PRs get merged I'll post PRs to update the other u16s that could be u8s.

Off top of my head these are: `num_signatures` and `num_static_accounts`

**Comment by apfitzge** [comment]
rebased sinced CI was skipping and i couldn't figure out how to get it to re-run jobs...

**Comment by tao-stones** [comment]
> rebased sinced CI was skipping and i couldn't figure out how to get it to re-run jobs...

Sometime I need to re-login to buildkite to make the `rerun` button available. Perhaps you were in similar situation

**Comment by apfitzge** [comment]
> Sometime I need to re-login to buildkite to make the rerun button available. Perhaps you were in similar situation

Ah no, it was the github-side CI tasks. Usually there's a re-run button, but there wasn't 🤷‍♂️

**Review by LucasSte — APPROVED** [review]

**Pull request merged**